### ISSUE TITLE
Legacy and Split UPE flows: setup intent ID is not attached to order meta during process_payment

### ISSUE THREAD

**Issue by htdat** [Automattic/woocommerce-payments#6576]
### Describe the bug

Similar to #6575, but intent_id order meta is not attached to order during process_payment. I am just separating the issue. 

### To Reproduce

Similar to #6575, purchase a subscription with a free trial to see this issue.

0. Enable legacy or split UPE flow.
1. Use WCPay Subscription and create a subscription product with free trial.
2. Purchase this product, and check out with this 3DS card `4000 0027 6000 3184`.
3. DO NOT authorize this payment in the pop-up iframe.
4. Check order meta of the newly created order in WP Admin > WooCommerce > Orders.
5. Issue: this setup intent ID is not saved in the order meta `_intent_id`.

### Actual behavior

intent_id for setup intents would be attached in the `process_payment`, that's similar to payment intents in these UPE flows, as well as setup intents in the traditional (card) flow.

### Additional context

On-going discussion: p1687513417167319-slack-C05AW31EKFF

**Comment by timur27** [comment]
I think we need to treat this issue blocked by https://github.com/Automattic/woocommerce-payments/issues/6575. Once the blocker is completed, we'll have the updated setup intent and need to attach the updated data to the order.

[This code](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/payment-methods/class-upe-payment-gateway.php#L557-L563) should accomplish all the goals of this issue. However, it's placed inside [the payment intent processing flow](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/payment-methods/class-upe-payment-gateway.php#L475) and thus won't cover set up intent processing. It might be sufficient to move it outside of 

```php
if ( $payment_needed ) {
...
} else {
... 
} 
```
block. Once placed after this block, the order should be updated with all the latest intent details, regardless of intent type.

**Comment by FangedParakeet** [comment]
Is this issue necessarily blocked by #6575? When we reach the `process_payment` function [here](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/payment-methods/class-upe-payment-gateway.php#L453) for a free-trial subscription, our [`processCheckout` function](https://github.com/Automattic/woocommerce-payments/blob/develop/client/checkout/api/index.js#L565) will still pass the setup intent ID via the `wcpay_payment_intent_id` parameter and we'll collect it as usual over [here](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/payment-methods/class-upe-payment-gateway.php#L462).

If we add an `else` block for [this `if ( $payment_needed )` condition](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/payment-methods/class-upe-payment-gateway.php#L475) and include [this `order_service->attach_intent_info_to_order()` call](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/payment-methods/class-upe-payment-gateway.php#L562) using the relevant setup intent ID and order, with the `$charge_id` value as `null`, wouldn't we still be able to attach the setup intent ID created on checkout page load to the newly created order, even though the setup intent itself would still be missing metadata from Stripe's perspective? 

I'm suggesting something like the below...

```php
if ( $payment_intent_id ) {
	list( $user, $customer_id ) = $this->manage_customer_details_for_order( $order );
	if ( $payment_needed ) {
		// ...
	} else {
		$setup_intent = $this->payments_api_client->get_setup_intent( $payment_intent_id );
		$this->order_service->attach_intent_info_to_order( $order, $payment_intent_id, $setup_intent['status'], $setup_intent['payment_method'], $customer_id, null, $currency );
	}
}
```

AFAICT this appears to add the setup intent ID to the WC order meta when I follow the steps in this issue description. Let me know if I'm missing anything obvious!

**Comment by timur27** [comment]
I agree @FangedParakeet - I haven't tried it myself, but it's worth testing, and if it works, treating this as a temporary solution before https://github.com/Automattic/woocommerce-payments/issues/6575 is done.

**Comment by frosso** [comment]
Heisenberg to validate in the next sprint - we want to ensure it's not happening on the dUPE.

**Comment by gpressutto5** [comment]
I can't get the pop-up iframe when using that credit card number. When I try to place the order, I get the following error:

> No such setupintent: 'seti_1OS2BDR03O9UZoSUsmY3c70N'

The fix Samir suggested looks correct, but I couldn't test it because this error is thrown before that part. Because of a refactor, it needs to be updated to:

```php
... 
} else {
	$setup_intent_request = Get_Setup_Intention::create( $payment_intent_id );
	/** @var WC_Payments_API_Setup_Intention $setup_intent */ // phpcs:ignore Generic.Commenting.DocComment.MissingShort
	$setup_intent = $setup_intent_request->send();
	$this->order_service->attach_intent_info_to_order( $order, $payment_intent_id, $setup_intent->get_status(), $setup_intent->get_payment_method_id(), $customer_id, null, $currency );
}
```

@timur27 Could you check if you get this same error I do when using this card number to pay for a Subscription with Trial?

**Comment by timur27** [comment]
> Could you check if you get this same error I do when using this card number to pay for a Subscription with Trial?

@gpressutto5 It took me a while but I could reproduce it. I think you found a bug which was existing since the legacy gateway implementation. To reproduce

1. enable WooPay
2. add a trial subscription and checkout with a 3DS card

It's reproducible [with the legacy card gateway](https://github.com/Automattic/woocommerce-payments/assets/22319444/5cdebe6a-5675-49a7-b675-0f910d92411e) also. If there's a need to double-check this: 
1. checkout `6.6.0` tag on your local 
2. make `is_upe_legacy_enabled`, `is_upe_split_enabled` and `is_upe_deferred_intent_enabled` return `false` in `includes/class-wc-payments-features.php`
3. enable WooPay
4. add a trial subscription and checkout with a 3DS card

I think we need to create a new issue for this with the low/medium priority (given that it's a zero-day bug which haven't had any reports yet) and address this separately. 

For the purpose of the current issue, please disable WooPay.

**Comment by timur27** [comment]
Coming back to the current issue. I think we should re-consider whether it's still relevant, in other words check [this](https://github.com/Automattic/woocommerce-payments/issues/6576#issuecomment-1820508975). Here’s how I understand the implementation at the moment: 

1. The only used and single gateway now is deferred UPE gateway, which mostly mirrors the legacy gateway. The scenario outlined in this issue is not applicable to this gateway type ([the setup intent is attached to the meta](https://github.com/Automattic/woocommerce-payments/assets/22319444/4c8f0dd1-09cc-4288-82bb-6f07361943d0)). It would be great to double-check this point to confirm if I'm not the only one who sees the correct behavior. 
2. The fix suggested [here](https://github.com/Automattic/woocommerce-payments/issues/6576#issuecomment-1608533991) is applied to the code which will be removed, because the branch 

```php 
if ( $payment_intent_id ) {

}
```
is only for the use case when the intent was created previously (legacy UPE & split UPE) and is being updated. 


Having said that, my understanding is this issue needs to be closed without further action. Let me know if the above reasoning in (2) makes sense and if you can re-test and confirm my statement in (1). If any further alignment will be needed afterward or if there's any misunderstanding from my side, I'm happy to align.

**Comment by gpressutto5** [comment]
Thanks, Timur! Since the bug is in a part of the code that will be removed in the next version, it makes sense to close this issue unless a patch for an older version is required.
